### PR TITLE
PWN-9270 - Refactor UiKitEditText for reuse

### PR DESCRIPTION
**Pull request by gslevinkov**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-9270

## Description of Work

- now can set icon instead of just ic_arrow
- extract style attrs init to class

## Screenshots (Optional)
Screenshot 2023-07-20 at 18 12 35

### CODE REVIEW

File: ui-kit/src/main/res/layout/widget_ui_kit_edittext.xml [hunk 1]
```diff
@@ -62,48 +69,39 @@
...
             app:layout_constraintBottom_toBottomOf="parent"
             app:layout_constraintEnd_toEndOf="parent"
             app:layout_constraintTop_toTopOf="parent"
-            android:contentDescription="@null" />
+            tools:src="@drawable/ic_close"
+            tools:visibility="visible" />
 
     </androidx.constraintlayout.widget.ConstraintLayout>
 
-    <TextView
```
Comment (gslevinkov) [inline ui-kit/src/main/res/layout/widget_ui_kit_edittext.xml:83]: was not used

File: ui-kit/src/main/kotlin/org/p2p/uikit/UiKitEditTextStyleAttrsApplier.kt [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+import androidx.core.content.res.getDrawableOrThrow
+import androidx.core.content.res.getIntOrThrow
+import androidx.core.content.res.getResourceIdOrThrow
+import androidx.core.view.isVisible
+import android.content.res.TypedArray
+import android.text.method.DigitsKeyListener
+import org.p2p.uikit.databinding.WidgetUiKitEdittextBinding
+
+internal class UiKitEditTextStyleAttrsApplier {
```
Comment (eduardmaximovich) [inline ui-kit/src/main/kotlin/org/p2p/uikit/UiKitEditTextStyleAttrsApplier.kt:11]: it's a pitty that [airbnb/paris](https://github.com/airbnb/paris) is not usable anymore with the new gradle and android policy

**Review by eduardmaximovich — APPROVED** [review]

File: ui-kit/src/main/kotlin/org/p2p/uikit/components/edittext/UiKitEditText.kt [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import org.p2p.uikit.utils.focusAndShowKeyboard
+import org.p2p.uikit.utils.inflateViewBinding
+
+class UiKitEditText @JvmOverloads constructor(
+    context: Context,
+    attrs: AttributeSet? = null,
+) : ConstraintLayout(context, attrs) {
+    private val binding: WidgetUiKitEdittextBinding by lazy(LazyThreadSafetyMode.NONE, ::inflateViewBinding)
+
+    val values: UiKitEditTextValues = UiKitEditTextValues(binding)
```
Comment (gslevinkov) [inline ui-kit/src/main/kotlin/org/p2p/uikit/components/edittext/UiKitEditText.kt:19]: now UiKitEditText consists of 2 parts:
1) `values` - get immutable values from this view
2) `mutator` - set new listeners / texts / tags and other stuff

all this was for less garbage in a whole class, now the code that modifies the view is in a separate class
you also have a visible contract what can you do with this view

Comment (gslevinkov) [reply ui-kit/src/main/kotlin/org/p2p/uikit/components/edittext/UiKitEditText.kt:19]: before that, this class was like this everything mixed with anything else
Screenshot 2023-07-23 at 17 22 09
Screenshot 2023-07-23 at 17 21 59

File: ui-kit/src/main/kotlin/org/p2p/uikit/components/edittext/UiKitEditTextStyleAttrsApplier.kt [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+import android.content.res.TypedArray
+import android.graphics.drawable.Drawable
+import android.text.method.DigitsKeyListener
+import org.p2p.uikit.R
+import org.p2p.uikit.databinding.WidgetUiKitEdittextBinding
+import org.p2p.uikit.utils.withImageOrGone
+import org.p2p.uikit.utils.withTextOrGone
+
+internal class UiKitEditTextStyleAttrsApplier {
+    private companion object {
```
Comment (gslevinkov) [inline ui-kit/src/main/kotlin/org/p2p/uikit/components/edittext/UiKitEditTextStyleAttrsApplier.kt:18]: all the available attrs that are REALLY working can be found here

i removed all the attributes connected to android: namespace to be more explicit.

File: ui-kit/src/main/kotlin/org/p2p/uikit/components/edittext/v2/NewUiKitEditText.kt [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+import android.content.Context
+import android.os.Parcelable
+import android.util.AttributeSet
+import org.p2p.uikit.R
+import org.p2p.uikit.components.edittext.UiKitEditTextSavedState
+import org.p2p.uikit.databinding.WidgetUiKitEditTextNewBinding
+import org.p2p.uikit.utils.focusAndShowKeyboard
+import org.p2p.uikit.utils.inflateViewBinding
+
+class NewUiKitEditText @JvmOverloads constructor(
```
Comment (gslevinkov) [inline ui-kit/src/main/kotlin/org/p2p/uikit/components/edittext/v2/NewUiKitEditText.kt:14]: to remove breaking changes - there is a new version for UiKitEditText - scalable and adoptive straight from UiKIt.

File: ui-kit/src/main/kotlin/org/p2p/uikit/components/edittext/v2/NewUiKitEditTextDelegate.kt [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+@Suppress("UNUSED_PARAMETER")
+fun newUiKitEditTextDelegate(
+    inflateListener: ((financeBlock: NewUiKitEditText) -> Unit)? = null,
+    onBindListener: ((view: NewUiKitEditText, item: NewUiKitEditTextCellModel) -> Unit)? = null,
+): AdapterDelegate<List<AnyCellItem>> =
+    adapterDelegateViewBinding<NewUiKitEditTextCellModel, AnyCellItem, ItemUiKitEditTextBinding>(
+        viewBinding = { _, parent -> parent.inflateViewBinding(attachToRoot = false) },
+    ) {
+//        val view = binding.root
```
Comment (gslevinkov) [inline ui-kit/src/main/kotlin/org/p2p/uikit/components/edittext/v2/NewUiKitEditTextDelegate.kt:17]: i will uncomment this in next pr

File: ui-kit/src/main/kotlin/org/p2p/uikit/components/edittext/v2/NewUiKitEditTextStyleAttrsApplier.kt [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        private val ATTR_INPUT_DIGITS = R.styleable.NewUiKitEditText_android_digits
+        private val ATTR_NEXT_FOCUS_DOWN = R.styleable.NewUiKitEditText_android_nextFocusDown
+        private val ATTR_IME_OPTIONS = R.styleable.NewUiKitEditText_android_imeOptions
+    }
+
+    /**
+     * Apply attrs values and styles
+     * use xml-values if they are not set explicitly
+     */
+    fun applyToView(attrs: TypedArray, binding: WidgetUiKitEditTextNewBinding) = with(binding) {
```
Comment (gslevinkov) [inline ui-kit/src/main/kotlin/org/p2p/uikit/components/edittext/v2/NewUiKitEditTextStyleAttrsApplier.kt:39]: before: 
Screenshot 2023-07-23 at 17 21 59

File: ui-kit/src/main/kotlin/org/p2p/uikit/components/edittext/v2/NewUiKitEditTextStyleAttrsApplier.kt [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+        private val ATTR_END_DRAWABLE = R.styleable.NewUiKitEditText_inputEndDrawable
+
+        private val ATTR_INPUT_TEXT_COLOR = R.styleable.NewUiKitEditText_inputTextColor
+        private val ATTR_INPUT_TEXT_STYLE = R.styleable.NewUiKitEditText_inputTextAppearance
+        private val ATTR_INPUT_BACKGROUND_TINT = R.styleable.NewUiKitEditText_inputBackgroundTint
+
+        private val ATTR_IS_INPUT_ENABLED = R.styleable.NewUiKitEditText_android_enabled
+        private val ATTR_INPUT_TYPE = R.styleable.NewUiKitEditText_android_inputType
+        private val ATTR_INPUT_DIGITS = R.styleable.NewUiKitEditText_android_digits
+        private val ATTR_NEXT_FOCUS_DOWN = R.styleable.NewUiKitEditText_android_nextFocusDown
```
Comment (gslevinkov) [inline ui-kit/src/main/kotlin/org/p2p/uikit/components/edittext/v2/NewUiKitEditTextStyleAttrsApplier.kt:31]: new edit text has it's own attributes. i decided not to use android: namespace for you to understand what can you do with this edit text using XML and app: namespace

File: ui-kit/src/main/kotlin/org/p2p/uikit/components/edittext/v2/NewUiKitEditTextValues.kt [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+import androidx.core.view.isVisible
+import android.graphics.drawable.Drawable
+import android.widget.EditText
+import org.p2p.uikit.databinding.WidgetUiKitEditTextNewBinding
+
+/**
+ * All the values that can be fetched from this view
+ */
+class NewUiKitEditTextValues(
```
Comment (gslevinkov) [inline ui-kit/src/main/kotlin/org/p2p/uikit/components/edittext/v2/NewUiKitEditTextValues.kt:11]: getters for UiKitEditText
all the values that you can get

File: ui-kit/src/main/kotlin/org/p2p/uikit/components/edittext/v2/NewUiKitEditTextMutator.kt [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+import org.p2p.core.utils.emptyString
+import org.p2p.uikit.R
+import org.p2p.uikit.databinding.WidgetUiKitEditTextNewBinding
+import org.p2p.uikit.utils.SimpleTagTextWatcher
+import org.p2p.uikit.utils.getColor
+
+/**
+ * Allows to change state of the view by settings text / listeners and other stuff
+ */
+class NewUiKitEditTextMutator(
```
Comment (gslevinkov) [inline ui-kit/src/main/kotlin/org/p2p/uikit/components/edittext/v2/NewUiKitEditTextMutator.kt:22]: setters
you can use this class if you want to change (mutate) the state of the view.

File: ui-kit/src/main/kotlin/org/p2p/uikit/components/edittext/v2/NewUiKitEditText.kt [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+    context: Context,
+    attrs: AttributeSet? = null,
+) : ConstraintLayout(context, attrs) {
+    private val binding: WidgetUiKitEditTextNewBinding by lazy(LazyThreadSafetyMode.NONE, ::inflateViewBinding)
+
+    val values: NewUiKitEditTextValues = NewUiKitEditTextValues(binding)
```
Comment (gslevinkov) [inline ui-kit/src/main/kotlin/org/p2p/uikit/components/edittext/v2/NewUiKitEditText.kt:20]: the only problem with this way of extracting code is that you violate the hollywood principle (https://deviq.com/principles/hollywood-principle). 

to get any value you need to call `binding.newEditText.values.inputText` 
to set listener or error `binding.newEditText.mutate().setError(...)`

I didn't want to make this view group big, that's why i separated set/get logic into a classes

**Review by 1sakov — APPROVED** [review]

**Pull request merged**